Update LSP6 specifications (#147)
lukso-network/LIPs: LSPs/LSP-6-KeyManager.md

## Specification

> **LSP6-KeyManager** interface id according to [ERC165]: `0xfb437414`.

Reviewer: Do we include ERC1271 function `isValidSignature` in the interface calculation?

At the moment it is not
<img width="621" alt="image" src="https://user-images.githubusercontent.com/31145285/205257744-ec79d3c0-a554-4d70-b79a-9e86ec794666.png">

But we use custom logic to check the `SIGN` permission within `isValidSignature`.

<img width="717" alt="image" src="https://user-images.githubusercontent.com/31145285/205257897-c2e2ddc6-d940-4471-ba3d-498f206793b8.png">

Imo, we should include `isValidSignature` in the interface ID calculation.

Reviewer: So far as mentioned by @frozeman we don't include the selector of `isValidSignature` because:

> The keymanager can also be used for non account smart contracts like NFTS, where it doesn't need to have the isValidsignature implemented. So i would rather keep it out.

Smart contracts implementing the LSP6 standard MUST implement the [ERC165] `supportsInterface(..)` function and MUST support the [ERC165], [ERC1271] and the LSP6 interface ids.

Every contract that supports the LSP6 standard SHOULD implement:

### Methods

#### target

```solidity
function target() external view returns (address)
```

Returns the `address` of the target smart contract controlled by this Key Manager. The controlled smart contract can be one of the following:
- ERC725X contract
- ERC725Y contract
- an ERC725 contract, implementing both ERC725X and ERC725Y (e.g: an [ERC725Account]).

#### isValidSignature

```solidity
function isValidSignature(bytes32 hash, bytes memory signature) external view returns (bytes4);
```

This function is part of the [ERC1271] specification, with additional requirements as follows:

> - MUST recover the address from the hash and the signature and return the [magic value] if the address recovered have the [**SIGN Permission**](#sign), if not, MUST return the [failure value].

Reviewer: ```suggestion
- MUST recover the address from the hash and the signature and return the [ERC1271 magic value] if the address recovered have the [**SIGN Permission**](#sign), if not, MUST return the [ERC1271 fail value].
```

#### getNonce

```solidity
function getNonce(address signer, uint128 channel) external view returns (uint256)
```

> Returns the nonce that needs to be signed by an allowed key to be passed into the [executeRelayCall](#executeRelayCall) functionS. A signer can choose his channel number arbitrarily.

Reviewer: ```suggestion
Returns the latest nonce for a signer on a specific channel. A signer can choose a channel number arbitrarily and use this nonce to sign a payload that can be executed as a meta-transaction by any address via [executeRelayCall](#executeRelayCall)  function.
```

_Parameters:_

- `signer`: the address of the signer of the transaction.
- `channel` :  the channel which the signer wants to use for executing the transaction.

_Returns:_ `uint256` , the current nonce.

> Payloads signed with nonces of the same channnel, needs to be executed by order. e.g, in channel X, payload signed with the second nonce will not be successfully executed until the payload signed with the first nonce is executed.

Reviewer: ```suggestion
Payloads signed with incremental nonces on the same channel for the same signer are executed in order. e.g, in channel X, the payload nb two signed with the second nonce will not be successfully executed until the payload nb one signed with the first nonce has been executed.
```

> Payloads signed with nonces of different channels are execution independant from each other. e.g, payload signed with the forth nonce of channel X can be successfully executed even before the execution of the payload signed with the first nonce of channel Y. 

Reviewer: ```suggestion
Payloads signed with nonces on different channels are executed independently from each other, regardless of when they got executed or if they got executed successfully or not. e.g, the payload signed with the fourth nonce on channel X can be successfully executed even if the payload signed with the first nonce of channel Y:
- was executed before.
- was executed and reverted.
```

> X and Y can be any arbitrary number between 0 and 2^128.

Read [what are multi-channel nonces](#what-are-multi-channel-nonces).

#### execute

```solidity

> function execute(bytes memory _calldata) external payable returns (bytes memory)

Reviewer: ```suggestion
function execute(bytes memory payload) external payable returns (bytes memory)
```

```

> Execute a payload on a set [ERC725 X or Y smart contract].

Reviewer: ```suggestion
Execute a payload on the linked [ERC725 X or Y smart contract].
```

MUST fire the [Executed event](#executed).

_Parameters:_

> - `_calldata`: The call data to be executed.

Reviewer: ```suggestion
- `payload`: The abi-encoded function call to be executed on the linked target contract.
```

_Returns:_ `bytes` , the returned data as abi-decoded bytes of the call on ERC725 smart contract, if the call succeeded, otherwise revert with a reason-string. 

_Requirements:_

> - The first 4 bytes of the `_calldata` payload MUST correspond to one of the function selector on the ERC725 smart contract such as:

Reviewer: ```suggestion
- The first 4 bytes of the `payload` payload MUST correspond to one of the function selector on the ERC725 smart contract such as:
```

    - [`setData(bytes32,bytes)`](./LSP-0-ERC725Account.md#setdata)
    - [`setData(bytes32[],bytes[])`](./LSP-0-ERC725Account.md#setdata-array)
    - [`execute(uint256,address,uint256,bytes)`](./LSP-0-ERC725Account.md#execute)
    - [`transferOwnership(address)`](./LSP-0-ERC725Account.md#transferownership)
    - [`acceptOwnership()`](./LSP-0-ERC725Account.md#acceptownership)

> - MUST send the value passed by the caller to the call on the ERC725 contracts. 

Reviewer: ```suggestion
- MUST send the value passed by the caller to the call on the linked target. 
```

> Non payable functions will revert in case of calling them and passing value along the call.

- The caller MUST have **permission** for the action being executed. Check [Permissions](#permissions) to know more.

#### execute (Array)

```solidity

> function execute(uint256[] memory values, bytes memory _calldata[]) external payable returns (bytes[] memory)

Reviewer: ```suggestion
function execute(uint256[] memory values, bytes memory payloads[]) external payable returns (bytes[] memory)
```

```

> Execute a batch of payloads on a set [ERC725 X or Y smart contract].

Reviewer: We can always put an anchore link to the `target` method. So people can see which contract we are executing to.

```suggestion
Execute a batch of payloads on the linked [target](#target) contract.
```

MUST fire the [Executed event](#executed) on each iteration.

_Parameters:_

- `values`: The array of values to be sent to the target contract along the call on each iteration. 

> - `_calldata`: The array of calldata to be executed on the target contract on each iteration.

Reviewer: ```suggestion
- `payloads`: The array of calldata payloads to be executed on the target contract on each iteration.
```

> _Returns:_ `bytes[]` , an array of returned data as array of abi-decoded bytes of the call on ERC725 smart contract, if the calls succeeded, otherwise revert with a reason-string. 

Reviewer: ```suggestion
_Returns:_ `bytes[]` , an array of returned data as abi-decoded array of `bytes[]` of the call on ERC725 smart contract, if the calls succeeded, otherwise revert with a reason-string. 
```

_Requirements:_

- The parameters length MUST be equal.

- The sum of each element of the `values` array MUST be equal to the value sent to the function.

- MUST comply to the requirements of the [`execute(bytes)`](#execute) function. 

#### executeRelayCall

```solidity

> function executeRelayCall(bytes memory signature, uint256 nonce, bytes memory _calldata) external payable returns (bytes memory)

Reviewer: ```suggestion
function executeRelayCall(bytes memory signature, uint256 nonce, bytes memory payload) external payable returns (bytes memory)
```

```

> Allows anybody to execute `_calldata` payload on a set [ERC725 X or Y smart contract], given they have a valid signature, specific to the payload passed, from permissioned controller.

Reviewer: ```suggestion
Allows anybody to execute a `payload` on the linked [target](#target) contract, given they have a valid signature, specific to the payload passed, from a permissioned controller.
```

MUST fire the [Executed event](#executed).

_Parameters:_
- `signature`: bytes65 ethereum signature.
- `nonce`: MUST be the nonce of the address that signed the message. This can be obtained via the `getNonce(address address, uint256 channel)` function.

> - `_calldata`: The call data to be executed.

Reviewer: ```suggestion
- `payload`: The calldata payload to be executed.
```

_Returns:_ `bytes` , the returned data as abi-decoded bytes of the call on ERC725 smart contract, if the call succeeded, otherwise revert with a reason-string. 

_Requirements:_

> - The address recovered from the signature and the digest signed MUST have **permission** for the action being executed. Check [Permissions](#permissions) to know more.

Reviewer: ```suggestion
- The address recovered from the signature and the digest signed MUST have **permission(s)** for the action(s) being executed. Check [Permissions](#permissions) to know more.
```

The digest signed MUST be constructed according to the [version 0 of EIP-191] with the following format:

```
0x19 <0x00> <KeyManager address> <LSP6_VERSION> <chainId> <nonce> <value> <payload>
```

    - `0x19`: byte intended to ensure that the `signed_data` is not valid RLP.
    - `0x00`: version 0 of the EIP191.
    - `KeyManager address`: The address of the Key Manager executing the payload.

>     - `LSP6_VERSION`: Version relative to the LSP6KeyManager defined as uint256 equal to 6.

Reviewer: ```suggestion
    - `LSP6_VERSION`: Version relative to the LSP6KeyManager defined as a uint256 equal to 6.
```

>     - `chainId`: The chainId of the blockchain where the Key Manager is deployed.

Reviewer: ```suggestion
    - `chainId`: The chainId of the blockchain where the Key Manager is deployed, as a uint256.
```

>     - `nonce`: The nonce to sign the payload with.

Reviewer: ```suggestion
    - `nonce`: The nonce to sign the payload with, as a uint256.
```

    - `payload`: The payload to be executed.

These parameters MUST be packed encoded (not zero padded, leading `0`s are removed), then hashed with keccak256 to produce the digest.

For signing, permissioned users should apply the same steps and sign the final hash got at the end.

- The nonce passed to the function MUST be a valid nonce according to the [multi-channel nonce](#what-are-multi-channel-nonces) section.

> - MUST send the value passed by the caller to the call on the ERC725 contracts. 

Reviewer: ```suggestion
- MUST send the value passed by the caller to the call on the linked target contract. 
```

> Non payable functions will revert in case of calling them and passing value along the call.

#### executeRelayCall (Array)

```solidity

> function executeRelayCall(bytes[] memory signatures, uint256[] memory nonces, uint256[] memory values, bytes[] memory _calldatas) external payable returns (bytes[] memory)

Reviewer: ```suggestion
function executeRelayCall(bytes[] memory signatures, uint256[] memory nonces, uint256[] memory values, bytes[] memory payloads) external payable returns (bytes[] memory)
```

```

> Allows anybody to execute a batch of `_calldatas` payloads on a set [ERC725 X or Y smart contract], given they have valid signatures, specific to the payloads passed, from permissioned controller/es.

Reviewer: ```suggestion
Allows anybody to execute a batch of `payloads` on the linked [target](#target) contract, given they have valid signatures specific to the payloads signed by permissioned controllers.
```

MUST fire the [Executed event](#executed) on each iteration.

_Parameters:_

- `signatures`: The array of bytes65 ethereum signature.
- `nonce`: The array of the nonce of the address/es that signed the digests. This can be obtained via the `getNonce(address address, uint256 channel)` function.
- `values`: The array of values to be sent to the target contract along the call on each iteration. 

> - `_calldata`: The array of call data to be executed.

Reviewer: ```suggestion
- `payloads`: The array of calldata payloads to be executed.
```

> _Returns:_ `bytes[]` , an array of returned data as array of abi-decoded bytes of the call on ERC725 smart contract, if the calls succeeded, otherwise revert with a reason-string. 

Reviewer: ```suggestion
_Returns:_ `bytes[]` , an array of returned as abi-decoded array of `bytes[]` of the linked target contract, if the calls succeeded, otherwise revert with a reason-string. 
```

_Requirements:_

- The parameters length MUST be equal.

- The sum of each element of the `values` array MUST be equal to the value sent to the function.

- MUST comply to the requirements of the [`executeRelayCall(bytes,uint256,bytes)`](#executerelaycall) function.

### Events

#### Executed

```solidity
event Executed(bytes4 indexed selector, uint256 indexed value);
```

MUST be fired when a transaction was successfully executed.

> The first parameter `selector` in the `Executed` event corresponds to the `bytes4` selector of the function being executed in the linked [target](#target).

Reviewer: perfect 👌

### Permissions

> The permissions MUST be checked against the caller in case of [`execute(bytes)`](#execute) and [`execute(uint256[],bytes[])`](#execute-array) functions. In case of [`executeRelayCall(bytes,uint256,bytes)`](#executerelaycall) and [`executeRelayCall(bytes[],uint256[],uint256[],bytes[])`](#executerelaycall-array), permissions should be checked against the signer address, recovered from the signature and the digest.

Reviewer: Maybe that would make it clearer with bullet points here?

----

The permissions MUST be checked against the following address, depending on the function/method being called:

- against the **caller** in the cases of [`execute(bytes)`](#execute) and [`execute(uint256[],bytes[])`](#execute-array): permissions MUST be checked  functions.
- against the **signer address**, recovered from the signature and the digest, in the cases of [`executeRelayCall(bytes,uint256,bytes)`](#executerelaycall) and [`executeRelayCall(bytes[],uint256[],uint256[],bytes[])`](#executerelaycall-array).

> The permissions MUST be stored as BitArray under the [`AddressPermissions:Permissions:<address>`](#addresspermissionspermissionsaddress) data key on the target.

Reviewer: add link to BitArray in LSP2.

> 

Reviewer: Let's add an additional paragraph to define the term "controller address".

```
In the descriptions of each permissions below, the term _"controller address"_ refers to an address that has some permissions set under the target contract linked to this Key Manager.
```

#### `ADDPERMISSIONS`

BitArray representation: `0x0000000000000000000000000000000000000000000000000000000000000002`

- Allows incrementing the length of the [`AddressPermissions[]`](#addresspermissions) data key and adding an address at a new index of the array.

- Allows adding permissions for a new controller key under the [`AddressPermissions:Permissions:<address>`](#addresspermissionspermissionsaddress) data key.

- Allows adding the [restrictions](#) for the [CALL](#call) and [SETDATA](#setdata) permissions stored respectively under the [`AddressPermissions:AllowedCalls:<address>`](#addresspermissionsallowedcallsaddress)  and [`AddressPermissions:AllowedERC725YDataKeys:<address>`](#addresspermissionsallowederc725ydatakeysaddress).

The value of these data keys SHOULD be validated before being set to avoid edge cases.

> All the actions above are done using the setData function

#### `CHANGEPERMISSIONS`

BitArray representation: `0x0000000000000000000000000000000000000000000000000000000000000004`

- Allows decrementing the length of the [`AddressPermissions[]`](#addresspermissions) data key and changing the address at an existing index of the array.

> - Allows changing permissions for an existing controller key under the [`AddressPermissions:Permissions:<address>`](#addresspermissionspermissionsaddress) data key.

Reviewer: let's use the term `address` alongside controller to avoid confusions:

```suggestion
- Allows changing permissions for an existing controller address under the [`AddressPermissions:Permissions:<address>`](#addresspermissionspermissionsaddress) data key.
```

> - Allows changing the existing [restrictions] for the [CALL](#call) and [SETDATA](#setdata) permissions stored respectively under the [`AddressPermissions:AllowedCalls:<address>`](#addresspermissionsallowedcallsaddress)  and [`AddressPermissions:AllowedERC725YDataKeys:<address>`](#addresspermissionsallowederc725ydatakeysaddress).

Reviewer: It's not necessarily the `CALL` permission here, it can be `STATICCALL` as well I think?
